### PR TITLE
adds support for line graphs, bar graphs, and pie graphs for basic graph components

### LINKED ISSUE #3: Charts Components
## 🚀 Feature Proposal

Add components for the following: 

- [ ] Bar Graph
- [ ] Line Graph
- [ ] Pie Chart

Charts should be based on [chart.js](https://www.chartjs.org/). 

## Motivation

Charts and graphs are an integral part of displaying data to the end user. Having a graph components will reduce development time and maintain consistency between graphs. 

## Example

HospitalRun/hospitalrun-frontend#387 requested a dashboard feature. Graphs/Charts could help in displaying data to the user in a dashboard.

### PR DESCRIPTION
**Pull request by jackcmeyer**
Fixes #3 .

**Changes proposed in this pull request:**
- adds bar graph, line graph, pie graph
- supports multiple datasets
- supports changing data colors
- supports changing title formatting
- supports legends
- supports axis label and customization

*Note: There may be more need for customization options and other features/functionalities of these graph components once use cases are more clearly defined*

Dependencies added: 
* [chart.js](https://bundlephobia.com/result?p=chart.js@2.8.0)
* [moment](https://bundlephobia.com/result?p=moment@2.24.0)

cc @HospitalRun/core-maintainers

### CODE REVIEW

**Comment by fox1t** [comment]
Hi @jackcmeyer can you pleas add [bundlephobia](https://bundlephobia.com/) link for all the deps you added?

**Comment by fox1t** [comment]
@jackcmeyer thanks for adding bundlephobia. I just added storybook to the project, can you please provide all use cases to document it.

**Comment by jackcmeyer** [comment]
@fox1t storybook update has been made. Also changed some props to be optional.

File: src/components/Graph/interfaces.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
+export interface Data {
```
Comment (fox1t) [inline src/components/Graph/interfaces.tsx]: This file has wrong extension: change it to `.ts`
tsx is just for components, we will add this to contribution markdown

File: test/linegraph.test.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
+import { LineGraph } from '../src/components/Graph'
+
+describe('it', () => {
+  it('renders without crashing', () => {
```
Comment (fox1t) [inline test/linegraph.test.tsx]: we need to use real descriptions for tests :)

File: test/piegraph.test.tsx [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
+import { PieGraph } from '../src/components/Graph'
+
+describe('it', () => {
```
Comment (fox1t) [inline test/piegraph.test.tsx]: We need real description on all tests

File: test/bargraph.test.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import * as React from 'react'
+import * as ReactDOM from 'react-dom'
+
+import { BarGraph } from '../src/components/Graph'
+
+describe('it', () => {
+  it('renders without crashing', () => {
```
Comment (fox1t) [inline test/bargraph.test.tsx]: We need real description on all tests

File: src/components/Graph/index.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        scales = {
+          xAxes: getAxes(yAxes, stacked),
+          yAxes: getAxes(xAxes, stacked),
+        }
+      }
+
+      config.options.scales = scales
+    }
+
+    this.graph = new ChartJs(this.chart, config)
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: I don't understand how you use `this.chart`: where you initialize it?

File: src/components/Graph/index.tsx [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+interface IBarGraph extends Graph, HasAxes {
+  /** Determines if the bar graph should be displayed in a horizontal manner */
+  horizontal?: boolean
+  /** Determines if the bar graph should be displayed as a stacked bar graph */
+  stacked?: boolean
+}
+
+class BarGraph extends Component<IBarGraph, {}> {
+  graph: any
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: Assign correct type here: `ChartJs`

File: src/components/Graph/index.tsx [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  chart: any
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: What is the chart type?

File: src/components/Graph/index.tsx [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+  }
+}
+
+interface IPieGraph extends Graph {
+  /** Determines if the pie graph should be displayed with a doughnut */
+  doughnut?: boolean
+}
+
+class PieGraph extends Component<IPieGraph, {}> {
+  graph: any
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: Same as above for both props

File: src/components/Graph/index.tsx [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+
+function getCommonChartConfigurations(type: string, graph: Graph): ChartConfiguration {
+  return {
+    type,
+    data: getChartData(graph),
+    options: getCommonChartOptions(graph),
+  }
+}
+
+interface IBarGraph extends Graph, HasAxes {
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: We will not use the `i` convention for Interfaces: you can drop it.

File: src/components/Graph/index.tsx [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+
+interface ILineGraph extends Graph, HasAxes {
+  /** Determines if the area under the line should be filled with the background color from the dataset */
+  fill?: boolean
+  /** Determines if the the Line Graphs should be displayed in a stacked manner */
+  stacked?: boolean
+}
+
+class LineGraph extends Component<ILineGraph, {}> {
+  graph: any
```
Comment (fox1t) [inline src/components/Graph/index.tsx]: Provide correct type: you can use `ChartJs`

File: test/piegraph.test.tsx (hunk 3 above)
Comment (tehKapa) [reply test/piegraph.test.tsx]: Done

File: src/components/Graph/index.tsx (hunk 5 above)
Comment (jackcmeyer) [reply src/components/Graph/index.tsx]: this.chart is used in the render method. 

The `ChartJs`requires an element to be passed in as the first argument to the constructor. this.chart is the element where the Chart should be rendered to. I've updated the type of `this.chart` to be `HTMLCanvasElement` so hopefully the code is a little more clear. 

This methodology is similar to how [react-chartjs-2](https://www.npmjs.com/package/react-chartjs-2#chartjs-instance) does it

**Comment by fox1t** [comment]
LGTM

**Pull request merged**